### ISSUE TITLE
Ошибка при оплате заказа "Order was not locked."

### ISSUE THREAD

**Issue by pro100roman24** [iiko/front.api.sdk#442]
Добрый день, столкнулись с проблемой.
Наш плагин вносит оплату в заказ и затем закрывает его.
Закрываем заказ методом:
PayOrderAndPayOutOnUser(IUser user, IBaseOrderEntity order, IPaymentType paymentType, Decimal sum, ChequeAdditionalInfo apiChequeAdditionalInfo)
Внесение оплаты происходит нормально, но при попытке закрыть заказ получаем ошибку:

ERROR [ 3] - Payment of order #5 (aea9157c-67e4-4c34-8603-bb2d0449395a) failed. Order was not locked.

Ошибку поймали на версии 8.4.6018.0
На предыдущих версиях никогда не сталкивались с такой ошибкой.

Подскажите из за чего могла произойти эта ошибка и как с ней справляться?
Это какая то новая особенность версии?

**Comment by pro100roman24** [comment]
В API логе такие записи:

[2023-04-27 12:22:27,202] DEBUG [58] - Plugin “PayOrder 3.0” is calling PayOrder operation
[2023-04-27 12:22:27,221]  INFO [58] - Remote payment of order #5 (aea9157c-67e4-4c34-8603-bb2d0449395a) is possible.
[2023-04-27 12:22:27,250] ERROR [58] - Payment of order #5 (aea9157c-67e4-4c34-8603-bb2d0449395a) failed. Order was not locked.
[2023-04-27 12:22:27,258] DEBUG [58] - Plugin “PayOrder 3.0” is calling AddErrorMessage operation

**Comment by pro100roman24** [comment]
> Добрый день. Спасибо за сигнал. Мы знаем об этой проблеме и даже завели на нее баг RMS-54589. Пока как временное решение могу предложить вставить паузу между добавлением в заказ оплаты и его оплатой. Либо переходите на оплату на локальном терминале ["Оплата заказа на текущей кассе"](https://iiko.github.io/front.api.doc/2023/02/22/pay-order-local.html). Там такой проблемы нет.

Сделал задержку в 1 секунду перед попыткой оплаты, теперь новая ошибка, которой тоже раньше никогда не встречал:

ERROR [14] - Payment of order #11 (3b216b8d-06c9-4b29-a4ca-41e54fcf3137) failed. Cannot use cafe session.

Лог плагина:
[2023-05-03 13:39:18,459]  INFO [14] - Предоплата на сумму 50 внесена в заказ 11. Заказ остался открытым.
[2023-05-03 13:39:18,489]  INFO [14] - В заказе 11 оплат на суммму: 50 
[2023-05-03 13:39:18,509]  INFO [14] - Суммы оплат (50) хватает на оплату заказа (50,00)
[2023-05-03 13:39:18,509]  INFO [14] - Пробуем оплатить и закрыть заказ
[2023-05-03 13:39:19,662] ERROR [14] - Payment of order #11 (3b216b8d-06c9-4b29-a4ca-41e54fcf3137) failed. Cannot use cafe session.

api.log:

[2023-05-03 13:39:17,896] DEBUG [59] - Plugin “PayOrder 3.0.1” is calling AuthenticateByPin operation
[2023-05-03 13:39:17,976] DEBUG [59] - Submitting edit session #-2147483647 for plugin “PayOrder 3.0.1” with 1 actions: AddExternalPaymentItem
[2023-05-03 13:39:18,196] DEBUG [59] - Updating iikocard discounts for orders: №11 (3b216b8d-06c9-4b29-a4ca-41e54fcf3137)
[2023-05-03 13:39:18,266]  WARN [59] - Couldn't update iikocard discounts for order №11 (3b216b8d-06c9-4b29-a4ca-41e54fcf3137)
[2023-05-03 13:39:18,307] DEBUG [59] - Submitted edit session #-2147483647 for plugin “PayOrder 3.0.1”
[2023-05-03 13:39:18,347] DEBUG [59] - Plugin “PayOrder 3.0.1” is calling AuthenticateByPin operation
[2023-05-03 13:39:18,367] DEBUG [59] - Submitting edit session #-2147483646 for plugin “PayOrder 3.0.1” with 1 actions: AddOrderExternalData
[2023-05-03 13:39:18,379] DEBUG [59] - Updating iikocard discounts for orders: №11 (3b216b8d-06c9-4b29-a4ca-41e54fcf3137)
[2023-05-03 13:39:18,439]  WARN [59] - Couldn't update iikocard discounts for order №11 (3b216b8d-06c9-4b29-a4ca-41e54fcf3137)
[2023-05-03 13:39:18,439] DEBUG [59] - Submitted edit session #-2147483646 for plugin “PayOrder 3.0.1”
[2023-05-03 13:39:18,509] DEBUG [59] - Plugin “PayOrder 3.0.1” is calling AuthenticateByPin operation
[2023-05-03 13:39:18,519] DEBUG [59] - Plugin “PayOrder 3.0.1” is calling SplitOrderBetweenCashRegisters operation
[2023-05-03 13:39:19,553] DEBUG [19] - Plugin “PayOrder 3.0.1” is calling PayOrder operation
[2023-05-03 13:39:19,623]  INFO [19] - Remote payment of order #11 (3b216b8d-06c9-4b29-a4ca-41e54fcf3137) is possible.
[2023-05-03 13:39:19,642] ERROR [19] - Payment of order #11 (3b216b8d-06c9-4b29-a4ca-41e54fcf3137) failed. Cannot use cafe session.
[2023-05-03 13:39:19,662] DEBUG [19] - Plugin “PayOrder 3.0.1” is calling AddErrorMessage operation

**Comment by pro100roman24** [comment]
Я так понимаю, что это все тот же баг, и вариант остается только переходить на оплату на локальном терминале?
Что влечет за собой переход на API V8

**Comment by letolica** [comment]
> [2023-05-03 13:39:19,642] ERROR [19] - Payment of order https://github.com/iiko/front.api.sdk/issues/11 (3b216b8d-06c9-4b29-a4ca-41e54fcf3137) failed. Cannot use cafe session.

Добрый день. Что в это время есть в `cash-server.log`?

**Comment by pro100roman24** [comment]
Вчера перевел наш плагин на API V8, добавил параметр isPaymentLocal = true  в метод оплаты, проблема прошла, заказы начали закрываться нормально.
Но на api V7 многие плагины работают ещё, желательно пофиксить эту проблему

**Comment by letolica** [comment]
Добрый день. Задача RMS-54589 в работе пока.

**Comment by pro100roman24** [comment]
@letolica  Подскажите, нет какого-то ещё решения? Или сроков по решению бага?

Дело в том, что сейчас временное решение - это оплата на локальном терминале, которая есть только в API v8
Но так как в эту версию апи сейчас постоянно вносятся изменения приходится делать новую сборку плагина под каждую версию айко. У меня уже итак есть сборка плагина под версии ниже 8.4.6 на v7, есть версия на v8 под 8.4.6-8.4.7, а сейчас ещё пришлось делать версию под 8.5.8, потому что там внесли правки в апи.
Это очень неудобно, хотелось бы собрать одну сборку на v7 и чтоб она работала на каком то ряде версий айко.

**Comment by letolica** [comment]
@pro100roman24, я очень понимаю вас. Все решения предложены в [комментарии](https://github.com/iiko/front.api.sdk/issues/442#issuecomment-1527620718). Вы пробовали вставить паузу между добавлением в заказ оплаты и его оплатой для плагинов на API.V7?
К сожалению, повлиять на скорость решения и дать прогнозы я не могу.
@rgalk, FYI.

**Comment by pro100roman24** [comment]
@letolica  да, я пробовал вставить задержку на одну секунду, ошибка Order was not locked е появлялась, но начинала появляться ошибка "Cannot use cafe session".
Поэтому этот workaround не работает, насколько я понял.

После перехода на v8 этой проблемы нет, просто собирать новую сборку плагина под каждую версию айки не совсем удобно мягко говоря.

А сегодня мы столкнулись с этой же ошибкой "Order was not locked" на версии 8.3.7
Но в api V8 от 8.3.7 в методе PayOrder ещё не было параметра isPaymentLocal, то есть нет возможности перейти на оплату на локальном терминале.
получается для 8.3.7 вообще нет никакого решения проблемы? Только обновлять iiko?
